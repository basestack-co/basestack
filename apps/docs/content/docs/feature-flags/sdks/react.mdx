---
title: React & NextJS
tags: Basestack, Feature Flags, React, SDK, Hooks, Components, State, Async, Flags, Flag, Initialize, Methods, Reference, Examples
---

import { File, Folder, Files } from 'fumadocs-ui/components/files';


Integration with [React ↗](https://react.dev/?ref=basestack.co) or [NextJS ↗](https://nextjs.org/?ref=basestack.co) helps and facilitates the process of testing and developing features in production and other environments.

## Getting started

For [React ↗](https://react.dev/?ref=basestack.co) and [NextJS ↗](https://nextjs.org/?ref=basestack.co), we chose not to create a separate SDK, as it would
essentially act as a wrapper for our [JavaScript SDK](/feature-flags/sdks/javascript). Instead, we’ll show you
how to fully use the JavaScript SDK within a React or Next.js project.
This approach ensures you can access all the features of a React SDK without
needing to install additional packages. One key advantage is that you retain
complete control over the code, allowing you to extend its functionality with
custom implementations.

#### Examples

- [React (Vite + TypeScript)](https://github.com/basestack-co/examples/tree/main/feature-flags/react-ts)
- [Next,js (Pages Router)](https://github.com/basestack-co/examples/tree/main/feature-flags/nextjs-pages-router)
- [Next,js (App Router)](https://github.com/basestack-co/examples/tree/main/feature-flags/nextjs-app-router)

### Install

First, let's install some packages!

```sh copy npm2yarn
npm install @basestack/flags-js
```

### Environment Variables

Here are the example environment variables needed for this guide. Feel free to rename them according to your project's requirements:

<Callout type="info" emoji="ℹ️">
  When it comes to environment variables, pay attention to the framework you're
  using. For example, `NEXT_PUBLIC_` is specific to Next.js, while in
  [Vite.js](https://vite.dev/?ref=basestack.co), it would be `VITE_`. Example
  `VITE_FEATURE_FLAGS_BASE_URL=` or Node.js `FEATURE_FLAGS_BASE_URL=`
</Callout>

```bash copy title=".env"
# BASESTACK FEATURE FLAGS
NEXT_PUBLIC_FEATURE_FLAGS_BASE_URL="https://flags.basestack.co/api/v1" # or your own URL
NEXT_PUBLIC_FEATURE_FLAGS_PROJECT_KEY=""
NEXT_PUBLIC_FEATURE_FLAGS_ENVIRONMENT_KEY=""
```

### Create a new folder called `feature-flags` in the `src/libs` directory

Inside the `feature-flags` folder, create the following files and folders:

<Files>
  <Folder name="src" defaultOpen>
    <Folder name="libs" defaultOpen>
      <Folder name="feature-flags" defaultOpen>
        <Folder name="hooks" defaultOpen>
          <File name="useFlag.tsx" />
          <File name="useFlags.tsx" />
          <File name="index.ts" />
        </Folder>
        <File name="index.tsx" />
        <File name="server.ts" />
      </Folder>
    </Folder>
  </Folder>
  <File name=".env" />
</Files>

Let's start by creating a Provider. In React, a Provider is a component that allows you to pass down data or state to all components within its tree without having to manually pass props at every level. This will enable you to use the hooks created below anywhere in your components.

In summary, this Provider initializes the `FlagsSDK, collects all the Feature Flags for a given environment to make them available in the context, and helps determine whether the Feature Flags have been initialized.

<Callout type="info" emoji="ℹ️">
  If you are only using React or Next.js in the Pages Router, you can remove the
  `“use client”` directive from the page.
</Callout>

```tsx copy title="src/libs/feature-flags/index.tsx"
"use client";

import React, { createContext, useEffect, useMemo, useState } from "react";
import { FlagsSDK, Flag, SDKConfig } from "@basestack/flags-js";

interface ContextState {
  client: FlagsSDK;
  isInitialized: boolean;
  flags: Flag[];
  error?: Error;
}

interface ProviderProps {
  children: React.ReactNode;
  config: SDKConfig;
  onSuccessfulInit?: (success: boolean) => void;
}

export const createFlagsClient = (config: SDKConfig) => {
  return new FlagsSDK({
    baseURL: config.baseURL,
    projectKey: config.projectKey,
    environmentKey: config.environmentKey,
  });
};

export const FeatureFlagsContext = createContext<ContextState | undefined>(
  undefined,
);

const FeatureFlagsProvider: React.FC<ProviderProps> = ({
  children,
  config,
  onSuccessfulInit,
}) => {
  const [state, setState] = useState<Omit<ContextState, "client">>({
    isInitialized: false,
    flags: [],
  });
  const client = useMemo(() => createFlagsClient(config), [config]);
  const value = useMemo(() => ({ ...state, client }), [state, client]);

  useEffect(() => {
    let isMounted = true;

    const initializeFlags = async () => {
      try {
        const response = await client.getAllFlags();

        if (isMounted) {
          setState((prev) => ({
            ...prev,
            flags: response.flags ?? [],
            isInitialized: true,
          }));
          onSuccessfulInit?.(true);
        }
      } catch (error) {
        if (isMounted) {
          setState((prev) => ({
            ...prev,
            error: error as Error,
            isInitialized: true,
          }));
          onSuccessfulInit?.(false);
        }
      }
    };

    initializeFlags();

    return () => {
      isMounted = false;
    };
  }, [client, onSuccessfulInit]);

  return (
    <FeatureFlagsContext.Provider value={value}>
      {children}
    </FeatureFlagsContext.Provider>
  );
};

export default FeatureFlagsProvider;
```

The `FlagsSDK` is versatile and can be used in various environments, including Node, Serverless environments, API Routes, Server Components, and API Endpoints. To avoid repeatedly initializing the client with environment variables, you can use the file below to simplify the process. Simply call this file as shown in the `usage` example below.

```ts copy title="src/libs/feature-flags/server.ts"
import { FlagsSDK, SDKConfig } from "@basestack/flags-js";

export class ServerFlagsSDK {
  private static instance: FlagsSDK;
  private static config: SDKConfig = {
    baseURL: process.env.NEXT_PUBLIC_FEATURE_FLAGS_BASE_URL,
    projectKey: process.env.NEXT_PUBLIC_FEATURE_FLAGS_PROJECT_KEY!,
    environmentKey: process.env.NEXT_PUBLIC_FEATURE_FLAGS_ENVIRONMENT_KEY!,
    // This is an example of how to use environment variables in Vite.js
    // environmentKey: import.meta.env.VITE_FEATURE_FLAGS_ENVIRONMENT_KEY!,
  };

  public static getInstance(): FlagsSDK {
    if (!ServerFlagsSDK.instance) {
      ServerFlagsSDK.instance = new FlagsSDK(ServerFlagsSDK.config);
    }
    return ServerFlagsSDK.instance;
  }
}
```

This hook, `useFlag`, is used to validate a Feature Flag by searching for a specific flag. It returns the necessary information for the specified flag. See the usage example below. You can customize or modify this hook to fit your requirements.

```tsx copy title="src/libs/feature-flags/hooks/useFlag.tsx"
"use client";

import { useContext, useEffect, useState, useMemo } from "react";
// Feature Flags Context
import { FeatureFlagsContext } from "@/libs/feature-flags";
import { Flag } from "@basestack/flags-js";

export const useFlag = <P = Record<string, unknown>,>(slug: string) => {
  const context = useContext(FeatureFlagsContext);
  const [flag, setFlag] = useState<Flag | null>(null);
  const [error, setError] = useState<Error | null>(null);

  if (!context) {
    throw new Error("useFlag must be used within a FeatureFlagsContext");
  }

  useEffect(() => {
    const fetchFlag = async () => {
      if (!context.isInitialized || !context.client) return;

      try {
        if (context.flags.length > 0) {
          const foundFlag = context.flags.find((f) => f.slug === slug);
          if (foundFlag) {
            setFlag(foundFlag);
          } else {
            setError(new Error(`Flag "${slug}" not found`));
          }
        } else {
          const fetchedFlag = await context.client.getFlag(slug);
          setFlag(fetchedFlag);
        }
      } catch (err) {
        setError(err instanceof Error ? err : new Error(String(err)));
      }
    };

    void fetchFlag();

    return () => {};
  }, [context.isInitialized, context.client, context.flags, slug]);

  return useMemo(
    () => ({
      ...flag,
      error,
      isInitialized: context.isInitialized,
      payload: (flag?.payload as P) ?? null,
    }),
    [error, context.isInitialized, flag],
  );
};
```

This hook, as the name suggests, is used to retrieve all the Feature Flags for a given environment. It can be used to display a list, perform new searches, or apply filtering.

```tsx copy title="src/libs/feature-flags/hooks/useFlags.tsx"
"use client";

import { useContext, useEffect, useState, useMemo } from "react";
// Feature Flags Context
import { FeatureFlagsContext } from "@/libs/feature-flags";
import { Flag } from "@basestack/flags-js";

interface FlagsState {
  flags: Flag[];
}

interface UseFlagsResult extends FlagsState {
  error: Error | null;
  isInitialized: boolean;
}

export const useFlags = (): UseFlagsResult => {
  const context = useContext(FeatureFlagsContext);
  const [{ flags }, setFlags] = useState<FlagsState>({ flags: [] });
  const [error, setError] = useState<Error | null>(null);

  if (!context) {
    throw new Error("useFlags must be used within a FeatureFlagsContext");
  }

  useEffect(() => {
    if (!context.isInitialized || !context.client) return;

    if (context.flags.length > 0) {
      setFlags({ flags: context.flags });
      return;
    }

    context.client
      .getAllFlags()
      .then((result) => setFlags(result))
      .catch((err) =>
        setError(err instanceof Error ? err : new Error(String(err))),
      );

    return () => {};
  }, [
    context.isInitialized,
    context.client,
    context.flags.length,
    context.flags,
  ]);

  return useMemo(
    () => ({
      error,
      isInitialized: context.isInitialized,
      flags,
    }),
    [flags, context.isInitialized, error],
  );
};
```

Finally, export all the created hooks to make them easy to import. For example: `import { useFlag, useFlags } from "@/libs/feature-flags/hooks";`

```ts copy title="src/libs/feature-flags/hooks/index.ts"
export * from "./useFlag";
export * from "./useFlags";
```

### Add the `FeatureFlagsProvider` to your App

The example below demonstrates how to add the `FeatureFlagsProvider` to your Next.js app using the App Router. It can also be used in the `_app.tsx` file or a React SPA app.

```tsx copy title="src/app/layout.tsx"
import React from "react";
import "./globals.css";
// Feature Flags
import FeatureFlagsProvider from "@/libs/feature-flags";

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body>
        <FeatureFlagsProvider
          config={{
            baseURL: process.env.NEXT_PUBLIC_FEATURE_FLAGS_BASE_URL,
            projectKey: process.env.NEXT_PUBLIC_FEATURE_FLAGS_PROJECT_KEY!,
            environmentKey:
              process.env.NEXT_PUBLIC_FEATURE_FLAGS_ENVIRONMENT_KEY!,
          }}
        >
          {children}
        </FeatureFlagsProvider>
      </body>
    </html>
  );
}
```

Your app is now fully equipped to take advantage of feature flags and other powerful functionalities offered by Basestack. To maximize the potential of Basestack, refer to the instructions provided in the supported hooks and components documentation.

### Usage

The examples demonstrate how to integrate the provided code into your project. If you find that an important example is missing, please open an issue at the following [link](https://github.com/basestack-co/basestack/issues).

#### React Component

Use hooks in React components to fetch a single Feature Flag or all Feature Flags.

To support types in the payload defined in the Dashboard within the _useFlag hook_, use the following syntax: `useFlag<YourPayloadType>`.

```tsx copy title="src/app/page.tsx"
"use client";

// Feature Flags Hooks
import { useFlag, useFlags } from "@/libs/feature-flags/hooks";

export default function Home() {
  const count = useFlag<{ text: string }>("count");
  const data = useFlags();

  return (
    <div>
      <main>
        {count.enabled && (
          <div>
            <a>
              {!count.enabled
                ? "Count is not enabled"
                : `${count?.payload?.text} 0`}
            </a>
          </div>
        )}

        <h3>All the available flags</h3>
        <ul>
          {data.flags.map((flag, index) => {
            return (
              <li
                key={index}
              >{`Slug: ${flag.slug} Enabled: ${flag.enabled}`}</li>
            );
          })}
        </ul>
      </main>
    </div>
  );
}
```

#### API Route Handler (App Router)

On the server side, use the _server file_ that already has the `FlagsSDK` initialized. Use this file wherever the server is needed, such as in API Routes, Server Components, API Endpoints, etc.

```ts copy title="src/app/api/count/route.ts"
import { ServerFlagsSDK } from "@/libs/feature-flags/server";

export async function GET() {
  const flagsClient = ServerFlagsSDK.getInstance();

  const flag = await flagsClient.getFlag("count");
  const data = await flagsClient.getAllFlags();

  return Response.json({ data, flag });
}
```

#### API Endpoint (Pages Router)

Here's an example of using the server instance in a Next.js Pages Router API endpoint:

```ts copy title="src/pages/api/hello.ts"
import type { NextApiRequest, NextApiResponse } from "next";
import { ServerFlagsSDK } from "@/libs/feature-flags/server";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse,
) {
  const flagsClient = ServerFlagsSDK.getInstance();

  const flag = await flagsClient.getFlag("count");
  const data = await flagsClient.getAllFlags();

  res.status(200).json({ data, flag });
}
```

#### Server Component

You can use the server instance in any server-side context, including `Server Components` in Next.js.

```tsx copy title="src/app/count/page.tsx"
import { ServerFlagsSDK } from "@/libs/feature-flags/server";

export default async function Page() {
  const flagsClient = ServerFlagsSDK.getInstance();
  const data = await flagsClient.getAllFlags();

  return (
    <div>
      <h3>All the available flags in server component</h3>
      <ul>
        {data.flags.map((flag, index) => {
          return (
            <li key={index}>{`Slug: ${flag.slug} Enabled: ${flag.enabled}`}</li>
          );
        })}
      </ul>
    </div>
  );
}
```

#### Direct Call with Context

You can access the client directly through the context, providing more control when working with useEffect, onClick handlers, or other scenarios where direct access is needed.

```tsx copy title="src/app/posts/page.tsx"
"use client";

import { useState, useEffect, useContext } from "react";
// Feature Flags Context
import { FeatureFlagsContext } from "@/libs/feature-flags";
import { Flag } from "@basestack/flags-js";

export default function Posts() {
  const [flags, setFlags] = useState<Flag[]>([]);
  const context = useContext(FeatureFlagsContext);

  useEffect(() => {
    async function fetchAllFlags() {
      const data = await context?.client.getAllFlags();
      setFlags(data?.flags ?? []);
    }
    fetchAllFlags();
  }, []);

  if (!flags) return <div>Loading...</div>;

  return (
    <ul>
      {flags.map((flag, index) => {
        return (
          <li key={index}>{`Slug: ${flag.slug} Enabled: ${flag.enabled}`}</li>
        );
      })}
    </ul>
  );
}
```

## Initialisation options

For details regarding the initialization options of @basestack/flags-js, please refer to the documentation related to the [Javascript SDK](/feature-flags/sdks/javascript#initialisation-options).

## Troubleshooting

### Why are my components rendering twice?

It's normal. Check if your App/Component is wrapped inside [React.StrictMode ↗](https://react.dev/reference/react/StrictMode?ref=basestack.co). The purpose of `React.StrictMode` is to help identify potential problems in your application by highlighting some unsafe practices or potential bugs during development.

When you wrap a component in `React.StrictMode`, it can cause some components to be rendered twice. This is expected behavior and is not specific to your code. Rendering twice is intentional and done by React on purpose.

```js copy title="main.ts"
const root = ReactDOM.createRoot(document.getElementById("root"));

root.render(
  <React.StrictMode>
    <FeatureFlagsProvider
      config={{
        baseURL: process.env.NEXT_PUBLIC_FEATURE_FLAGS_BASE_URL,
        projectKey: process.env.NEXT_PUBLIC_FEATURE_FLAGS_PROJECT_KEY!,
        environmentKey:
          process.env.NEXT_PUBLIC_FEATURE_FLAGS_ENVIRONMENT_KEY!,
      }}
    >
        <App />
    </FeatureFlagsProvider>
  </React.StrictMode>,
);
```
