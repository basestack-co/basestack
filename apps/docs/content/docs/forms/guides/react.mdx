---
title: NextJS & React
icon: Code
tags: Basestack Forms, NextJS, React Integration, Form submission, Fetch API, Axios, React, Next.js, API integration, JSON response, Web development, Developer tools, "2025"
---

import { Step, Steps } from 'fumadocs-ui/components/steps';

[React](https://react.dev/?ref=basestack.co) is a JavaScript library for building user interfaces, allowing developers to create interactive, dynamic web applications with reusable components. [Next.js](https://nextjs.org/?ref=basestack.co) is a React framework that enables server-side rendering, static site generation, and simplified routing for building fast, scalable web applications.

To get started, youâ€™ll need the following:

- A [Basestack Forms](https://forms.basestack.co) Account
- A React or Next.js project with a Form component

<Steps>

<Step>
### Use this on your React Project to send form data to your endpoint.

Send data using the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API?ref=basestack.co), [Axios](https://axios-http.com/?ref=basestack.co) or Any [HTTP](https://developer.mozilla.org/en-US/docs/Web/HTTP?ref=basestack.co) Client and ensure the mode on the query string is set to `mode=rest` for retrieving the response in JSON format.
</Step>
<Step>
### Copy and paste this into your React component.

Feel free to use the code below to send a test submission.


```tsx copy title="contact.tsx"
import React, { useState } from "react";

const FormComponent = () => {
  // You can also use rect-hook-form or formik for form handling
  // check out the docs for more info https://react-hook-form.com/
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [message, setMessage] = useState("");

  const onHandleSubmit = (e) => {
    e.preventDefault();
    e.stopPropagation();

    // You can also use axios or any other library to make the request
    // You can also use async await to make the request
    fetch("https://forms.basestack.co/api/v1/s/[KEY]?mode=rest", {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ name, email, message }),
    })
      .then((res) => res.json())
      .then((data) => {
        if (data.code === 200) {
          console.log("Form submitted successfully");
        }
      })
      .catch((error) => {
        console.log("Error submitting form", error);
      });
  };

  return (
    <form onSubmit={(e) => onHandleSubmit(e)}>
      <div>
        <label htmlFor="name">Name:</label>
        {/*  <!-- name each of your inputs as you wish --> */}
        <input
          type="text"
          id="name"
          name="name"
          required
          autoComplete="name"
          aria-label="Name"
          placeholder="Enter your name"
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
      </div>

      <div>
        <label htmlFor="email">Email:</label>
        <input
          type="email"
          id="email"
          name="email"
          required
          autoComplete="email"
          aria-label="Email"
          placeholder="Enter your email"
          value={email}
          onChange={(e) => setEmail(e.target.value)}
        />
      </div>

      <div>
        <label htmlFor="message">Message:</label>
        <textarea
          id="message"
          name="message"
          required
          aria-label="Message"
          placeholder="Enter your message"
          value={message}
          onChange={(e) => setMessage(e.target.value)}
        ></textarea>
      </div>

      {/*  <!-- your other form fields go here --> */}

      <button type="submit">Submit</button>
    </form>
  );
};

export default FormComponent;
```
</Step>
</Steps>
