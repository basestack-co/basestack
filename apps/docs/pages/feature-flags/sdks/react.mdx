---
title: React SDK Reference
description: Integration with React helps and facilitates the process of testing and developing features in production and other environments.
---

import { Steps, Callout } from "nextra/components";

# React SDK Reference

Integration with React helps and facilitates the process of testing and developing features in production and other environments. This integration consists of a set of plug-and-play components to accelerate the development process in your project.

## Getting started

<Steps>
  ### Install

First, let's install some packages!

```sh copy npm2yarn
npm i @basestack/flags-js-sdk @basestack/flags-react-sdk
```

### Create a client provider

In your index.js file, import FlagsProvider from @basestack/flags-react-sdk and include the configuration parameters specific to your Basestack project. You can find these parameter values on your project settings page.

```js copy filename="index.js"
import FlagsJS from "@basestack/flags-js-sdk";
import { FlagsProvider } from "@basestack/flags-react-sdk";

const sdk = new FlagsJS({
  apiUrl: "https://your-basestack-hosted-app-domain.com/api/v1",
  projectKey: "xxxxx",
  envKey: "xxxxx",
});

root.render(
  <FlagsProvider
    sdk={sdk}
    onSuccessfulInit={() => console.log("Successful Init FlagsJS SDK")}
  >
    <App />
  </FlagsProvider>,
);
```

Your app is now fully equipped to take advantage of feature flags and other powerful functionalities offered by Basestack. To maximize the potential of Basestack, refer to the instructions provided in the supported hooks and components documentation.

### Usage

```jsx copy filename="MyComponent.js"
import { useFlag } from "@basestack/flags-react-sdk";

const AppComponent = () => {
  const { enabled } = useFlag("private_msg_2");
  const privateChat = useFlag("private_chat");

  return (
    <div>
      {enabled && <div>This is a great feature</div>}
      {privateChat.enabled && <div>This is a great private chat</div>}
    </div>
  );
};
```

</Steps>

## Cache vs Async

The SDK provides two alternatives for fetching a feature flag: the **useFlag** and **useFlagAsync** hooks. Both serve the same purpose of retrieving a flag by its name. However, the 'useFlag' hook utilizes the cached version, resulting in faster access, while 'useFlagAsync' hook fetches the latest version directly from the server.

Which hook should you choose? It depends on your specific needs. If having the most up-to-date version of the flag is crucial and it changes frequently, then opt for the async hooks. On the other hand, if the flag doesn't undergo frequent changes and you prioritize performance, the hook with cache support are the ideal choice. Assess your requirements and select the method that aligns best with your priorities.

<Callout>
  The cache functions as a state container that automatically updates every time
  your app is restarted or updated.
</Callout>

```js copy filename="main.js"
import { useFlag, useFlagAsync } from "@basestack/flags-react-sdk";

const MyComponent = () => {
  // Get from the cache, no call for the server
  const cache = useFlag("header");
  // Get a new fresh version from the server
  const server = useFlagAsync("header");

  return (
    <div>
      {!server.isLoading && server.flag.enabled && (
        <div>
          This feature is waiting for the server response to check if we can
          enabled
        </div>
      )}
      {cache.enabled && <div>This is enabled feature from the cache</div>}
    </div>
  );
};
```

## Hooks

The library fully supports React hooks, providing the **useFlag**, **useFlagAsync**, **useFlags** and **useFlagsAsync** hook for a programmatic approach.

### useFlag & useFlagAsync Hooks

With **useFlag** or **useFlagAsync**, you can create and utilize multiple flags without any restrictions; simply change the names as needed. For practical implementation, refer to the examples provided.

```js copy
import { useFlag, useFlagAsync } from "@basestack/flags-react-sdk";

const MyComponent = () => {
  const { enabled } = useFlag("private_msg_2");
  const privateChat = useFlag("private_chat");
  // Get a new fresh version from the server
  const server = useFlagAsync("header");

  return (
    <div>
      {!server.isLoading && server.flag.enabled && (
        <div>
          This feature is waiting for the server response to check if we can
          enabled
        </div>
      )}
      {enabled && <div>This is a great feature</div>}
      {privateChat.enabled && <div>This is a great private chat</div>}
    </div>
  );
};
```

#### Remote Flags

Remote Flags share the same characteristics as regular Flags, but they come with an additional advantage – a data payload.

This payload empowers you to furnish real-time properties to your project's features. For instance, if certain features of a project require dynamic color changes, size adjustments, or any other data variations across different environments, utilizing the data payload of a Remote Flag becomes the ideal solution.

```js copy
import { useFlag, useFlagAsync } from "@basestack/flags-react-sdk";

const MyComponent = () => {
  const { enabled, payload } = useFlag("my_remote_flag");
  // Get a new fresh version from the server
  const server = useFlagAsync("header");

  return (
    <div>
      {!server.isLoading && server.flag.enabled && (
        <div style={{ color: server.flag.payload.color }}>
          This feature is waiting for the server response to check if we can
          enabled
        </div>
      )}
      {enabled && (
        <div style={{ color: payload.color }}>
          This is a great remote feature
        </div>
      )}
    </div>
  );
};
```

### useFlags & useFlagsAsync Hooks

For scenarios where you need to fetch all flags simultaneously, make use of the **useFlags** and **useFlagsAsync** hooks. These hooks enable efficient retrieval of multiple flags in a single operation.

```js copy
import { useFlags, useFlagsAsync } from "@basestack/flags-react-sdk";

const MyComponent = () => {
  const allFlags = useFlags();
  // Or use the useFlagsAsync to get fresh list from the server, this will update the cache state
  const allFlagsFromServer = useFlagsAsync();

  return (
    <div>
      {allFlags.map(({ enabled, slug }) => {
        if (!enabled) return null;

        return <p>{slug}</p>;
      })}
    </div>
  );
};
```

## Components

### Flag & FlagAsync Components

The pre-built components **Flag** or **FlagAsync** accepts one or multiple child components. Any components placed inside the **Flag** wrapper will only be rendered when the flag exists or enabled.

```js copy
import { Flag, FlagAsync } from "@basestack/flags-react-sdk";

const AppComponent = () => {
  return (
    <div>
      {/* Get a Flag from the cache */}
      <Flag name="private_msg_2">
        <YourFeature />
      </Flag>

      {/* Get a Flag from the cache and use as a Remote Flag. Ps. You can do the same with the FlagAsync Component */}
      <Flag
        name="my_remote_flag"
        render={(flag) => (
          <div style={{ color: flag.payload.color }}>
            A Remote flag inside a component
          </div>
        )}
      />

      {/* Get a Flag from the server */}
      <FlagAsync name="private_msg_2">
        <YourFeature />
      </FlagAsync>
    </div>
  );
};
```

## TypeScript Support

The SDK is built with TypeScript. Therefore, when utilized in a TypeScript project, it comes with basic type inference out of the box.

```js copy filename="main.ts"
import { useFlag } from "@basestack/flags-react-sdk";

const flag = useFlag("new_header_v2");

flag.enabled; // boolean
flag.createdAt; // Date
flag.slug; // string
flag.description; // string
// ...
```

#### Typescript Type Definition

```sh copy
interface Flag {
  slug: string;
  enabled: boolean;
  payload?: unknown;
  expiredAt?: Date;
  createdAt: Date;
  updatedAt: Date;
  description: string;
}
```

## SDK Reference

## Initialisation options

For details regarding the initialization options of @basestack/flags-js-sdk, please refer to the documentation related to the [Javascript SDK](/feature-flags/sdks/javascript#initialisation-options).

## Troubleshooting

### Why are my components rendering twice?

It's normal. Check if your App/Component is wrapped inside [React.StrictMode ↗](https://react.dev/docs/strict-mode). The purpose of `React.StrictMode` is to help identify potential problems in your application by highlighting some unsafe practices or potential bugs during development.

When you wrap a component in `React.StrictMode`, it can cause some components to be rendered twice. This is expected behavior and is not specific to your code. Rendering twice is intentional and done by React on purpose.

```js copy filename="main.ts"
const root = ReactDOM.createRoot(document.getElementById("root"));

root.render(
  <React.StrictMode>
    <FlagsProvider
      sdk={sdk}
      onSuccessfulInit={() => console.log("Successful Init FlagsJS SDK")}
    >
      <App />
    </FlagsProvider>
  </React.StrictMode>,
);
```

## Examples

**coming soon**
