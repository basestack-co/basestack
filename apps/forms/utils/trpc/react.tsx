"use client";

import React, { useState } from "react";
// Types
import { QueryClientProvider, type QueryClient } from "@tanstack/react-query";
import { type AppRouter } from "server/api/root";
// Server
import { loggerLink, unstable_httpBatchStreamLink } from "@trpc/client";
import { createTRPCReact } from "@trpc/react-query";
import { type inferRouterInputs, type inferRouterOutputs } from "@trpc/server";
// Utils
import SuperJSON from "superjson";
import { createQueryClient } from "./query-client";

let clientQueryClientSingleton: QueryClient | undefined = undefined;
const getQueryClient = () => {
  if (typeof window === "undefined") {
    // Server: always make a new query client
    return createQueryClient();
  }
  // Browser: use singleton pattern to keep the same query client
  return (clientQueryClientSingleton ??= createQueryClient());
};

export const api = createTRPCReact<AppRouter>();

export type RouterInputs = inferRouterInputs<AppRouter>;

export type RouterOutputs = inferRouterOutputs<AppRouter>;

export function TRPCReactProvider(props: { children: React.ReactNode }) {
  const queryClient = getQueryClient();

  const [trpcClient] = useState(() =>
    api.createClient({
      links: [
        loggerLink({
          enabled: (op) =>
            process.env.NODE_ENV === "development" ||
            (op.direction === "down" && op.result instanceof Error),
        }),
        unstable_httpBatchStreamLink({
          transformer: SuperJSON,
          url: getBaseUrl() + "/api/trpc",
          headers: () => {
            const headers = new Headers();
            headers.set("x-trpc-source", "nextjs-react");
            return headers;
          },
        }),
      ],
    }),
  );

  return (
    <QueryClientProvider client={queryClient}>
      {/*  @ts-expect-error - client prop Its working on Feature Flags, this started randomly failing */}
      <api.Provider client={trpcClient} queryClient={queryClient}>
        {props.children}
      </api.Provider>
    </QueryClientProvider>
  );
}

function getBaseUrl() {
  if (typeof window !== "undefined") return window.location.origin;
  if (process.env.VERCEL_URL) return `https://${process.env.VERCEL_URL}`;
  return `http://localhost:${process.env.PORT ?? 3000}`;
}
