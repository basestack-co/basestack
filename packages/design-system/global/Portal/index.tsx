import React, { useRef, useEffect, useState, JSX } from "react";
import { createPortal } from "react-dom";

interface PortalProps {
  children: React.ReactNode;
  selector: string;
}

const Portal = ({ children, selector }: PortalProps): JSX.Element => {
  const ref = useRef<any>(null);
  const [mounted, setMounted] = useState(false);

  useEffect(() => {
    ref.current = document.querySelector(selector);
    setMounted(true);
  }, [selector]);

  return <>{mounted ? createPortal(children, ref.current) : null}</>;
};

export default Portal;
