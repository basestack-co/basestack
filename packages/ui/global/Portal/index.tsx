import React, { useRef, useEffect, useState } from "react";
import { createPortal } from "react-dom";

interface PortalProps {
  children: React.ReactNode;
  selector: string;
}

export default function Portal({ children, selector }: PortalProps) {
  const ref = useRef<any>();
  const [mounted, setMounted] = useState(false);

  useEffect(() => {
    // @ts-ignore
    ref.current = document.querySelector(selector);
    setMounted(true);
  }, [selector]);

  return mounted ? createPortal(children, ref.current) : null;
}
