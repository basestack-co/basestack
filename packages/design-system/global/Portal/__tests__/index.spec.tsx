import { cleanup, render } from "@testing-library/react";
import Portal from "..";

afterEach(cleanup);

describe("Portal Component", () => {
  it("Portal renders correctly", () => {
    const { container } = render(
      <div id="portal">
        <Portal selector="#portal">Portal Content</Portal>
      </div>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
