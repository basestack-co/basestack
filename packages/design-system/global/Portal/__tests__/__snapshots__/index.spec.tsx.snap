// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal Component Portal renders correctly 1`] = `
<div
  id="portal"
>
  Portal Content
</div>
`;
